feat: bind client with router routes

Brentlok opened the pull request:
This PR includes refactored client types that bind router routes with client similarly to tRPC
I've left most of the logic as it was, I'm happy to help with further refactors

https://github.com/user-attachments/assets/b3a03fde-163f-4644-bcf8-59cea3f53aec

joschan21 commented:
woah that looks incredible, reviewing today

joschan21 commented:
wasn't quite able to get this working yesterday, could you share your `./src/server` where you import the client creator and jstack from (e.g. see video at 0:08)? The client types are correct, but seems like there is an issue with actually accessing the properties

![Screenshot 2025-04-04 at 10 05 00](https://github.com/user-attachments/assets/51fd2db1-973a-4566-b440-98a64b603771)

jcodog commented:
Same issue as @joschan21 so it would be beneficial if you could explain how you got it to work without this issue in your code. The changes appear to break how the client is being defined and works.

![image](https://github.com/user-attachments/assets/30ff697b-c24f-48ed-a35b-da357f6648f2)

joschan21 commented:
got it to work, hell yeah https://screen.studio/share/Uj8GEZYW

Brentlok commented:
This is everything that I have in my `src/server`

```ts
const j = jstack.init()

const publicProcedure = j.procedure

const postRouter = j.router({
    recent: publicProcedure.get(({ c }) => {
      return c.json({ title: "first post" })
    }),
    all: publicProcedure.get(({ c }) => {
      return c.json({ title: "first post" })
    }),
})

const testRouter = j.router({
  test: publicProcedure.get(({ c }) => {
    return c.json({ title: "first post" })
  }),
})

const api = j
  .router()
  .basePath("/api")
  .use(j.defaults.cors)
  .onError(j.defaults.errorHandler)

const appRouter = j.mergeRouters(api, {
    post: postRouter,
    test: testRouter,
})

type AppRouter = typeof appRouter

export const client = createClient<AppRouter>({
    baseUrl: '',
})
```

Brentlok commented:
It looks like when you pass only single router the `AppRouter` type is completely different than when you're passing multiple ones.
Single:
<img width="450" alt="image" src="https://github.com/user-attachments/assets/85a3ed97-bdb4-449b-9855-b9d2eb80d5a6" />
Two:
<img width="848" alt="image" src="https://github.com/user-attachments/assets/b5c0bbbb-defd-4c7f-b625-7d0ad6a5ce56" />

And I was testing only with two separate routers merged together. I'll address the issue 🙏

Brentlok commented:
I've updated the code. Now everything should work like a charm.
Let me explain how it works and what's the trick:
In order for our IDE to navigate (when clicking with CMD pressed) we need to use "original" keys when mapping newly created type. Simple example:

```ts
const routes = {
    register: 'register',
    list: 'list'
}

const mapRoutes = <T extends Record<string, any>>(routes: T) => {
    const generatedRoutes = {} as {
        [K in keyof T]: RouterEndpoint<T[K]>
    }
    // some magic

    return generatedRoutes
}
```

If we touch the key (keyof T) even slightly, it will break the connection between those two types.

To keep the origial object passed into `j.mergeRouters` without changing original logic I did the following:
I've created type symbol (we don't need a symbol in the runtime)

```ts
export declare const UnmergedRoutersSymbol: unique symbol
```

And then I've lied a little to a typescript and added this symbol property into return type from the result of `j.mergeRouters` function:

```ts
type MergedRouter<R extends RoutersToMerge> = Router<InferSchemaFromRouters<R>> & { [UnmergedRoutersSymbol]: R }
```

Finally In our `Client` type I can access this original hidden object with our symbol and map it into a new object

```ts
type GetUnmergedRouters<T> = T extends { [UnmergedRoutersSymbol]: infer S }
    ? S
    : never
```

Brentlok commented:
I hope that my explanation is clear enough for you guys, to be fair it's hard to explain some parts for me without writing long article 😅 

If you have any questions, or you've spotted another issue please let me know. I'm happy to help with them 🙏

joschan21 commented:
@Brentlok merged your commits in #72, appreciate if you take a look and have feedback. Super happy about how this works, thanks so much for the PR!
